core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory.java

```diff
@@ -186,23 +194,40 @@ public void handleNameResolutionError(Status error) {
 
     @Override
     public void handleSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
-      if (stateInfo.getState() == SHUTDOWN && stickinessState != null) {
+      Ref<ConnectivityStateInfo> stateInfoRef = getSubchannelStateInfoRef(subchannel);
+      ConnectivityState stateBefore = stateInfoRef.value.getState();
+      if (stateBefore == SHUTDOWN) {
+        return;
+      }
+      ConnectivityState newState = stateInfo.getState();
+      if (newState == SHUTDOWN && stickinessState != null) {
         stickinessState.remove(subchannel);
       }
       if (subchannels.get(subchannel.getAddresses()) != subchannel) {
         return;
       }
-      if (stateInfo.getState() == IDLE) {
+      if (newState == IDLE) {
         subchannel.requestConnection();
```

Comment:
@zhangkun83 isn't this also redundant?  When a new channel is created the connection is already requested.  What will calling it twice do?  Also, won't this prevent channels from ever going idle?